### PR TITLE
fix(ecma/transform/compat): arrow parameters in class field

### LINKED ISSUE #3471: Arrow functions as class properties compile incorrectly
### Describe the bug

When using specific browser target to reduce how much compilation is being applied to the code, an edge case surfaces that incorrectly transpiles arrow functions as class properties, losing the correct reference to `this`. See the attached playground and you'll notice `_this.getBar(b)` is referencing `_this` which is not a valid reference.

### Input code

```typescript
class Foo {
  getBar(b) {
    return b * 2
  }

  getFoo = ({ b } = {}) => {
    return this.getBar(b)
  }
}

const f = new Foo()
console.log(f.getFoo({ b: 2 }))
```

### Config

```js
{
    env: {
      targets: '> 0.25% and last 2 versions, not dead, not ie <= 11',
      coreJs: 3,
      mode: 'entry'
    },
    jsc: {
      parser: {
        jsx: true,
        exportDefaultFrom: true
      },
      transform: {}
    }
  }
}
```

### Playground link

https://play.swc.rs/?version=1.2.133&code=H4sIAAAAAAAAA0vOSSwuVnDLz1eo5lJQSE8tcUos0kjSBPMUFIpSS0qL8hSSFLQUjIACtVwQNSDltgoa1UCJWiCjulZTwdYOVUtJRmaxHtw4sF6g7uT8vOIShTSgnrzUcpCtGppgsfycVL2c%2FHSNND2I6SCjrRSMFGo1NQHVZOk3oQAAAA%3D%3D&config=H4sIAAAAAAAAA0WO3QqDMAyFXyUEdidjE9yFbHuDPUSwUTq0lSbKRPrua53Du5Pz85EV39JgveJIQThkJYtT%2BmCN3AwkTbCjYpFqyWqpF44FKoWONVekSlnvvfCeFjhYZ9slkxo%2FjIFFjohc1%2F%2BbMYEGb6ZsrKjLyBvwhvFg7Dsrr72oYUo3u3nbbG8kPD7hci6rE5Az0JMolDBzEOudFOC8gmEyP2UZ7g%2B4XjHGL6hpzCH9AAAA

### Expected behavior

this.getBar(b) call should work!

### Actual behavior

Error is thrown `Uncaught TypeError: _this.getBar is not a function`, because `_this` is not pointing to the class instance.

### Version

1.2.133

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Austaras**
**Related issue (if exists):**
closes #3471

### CODE REVIEW

File: crates/swc_ecma_transforms_compat/src/es2015/parameters.rs [hunk 1]
```diff
@@ -536,6 +564,29 @@ impl VisitMut for Params {
                             return_type: Default::default(),
                         },
                     });
+                    *e = match (self.in_prop, local_vars) {
+                        (true, Some(var_decl)) => Expr::Arrow(ArrowExpr {
```
Comment (kdy1) [inline crates/swc_ecma_transforms_compat/src/es2015/parameters.rs:568]: I think we should not generate an arrow function from here, as we apply `arrow` before `parameters`.

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!
I think we may change the pass ordering

File: crates/swc_ecma_transforms_compat/src/es2015/parameters.rs (hunk 1 above)
Comment (Austaras) [reply crates/swc_ecma_transforms_compat/src/es2015/parameters.rs:568]: A little bit context is needed. As discussed in #3268, Safari somehow still has bug for parameters until today(swc have `safari_id_destructuring_collision_in_function_expression` to fix it), so even if class properties is supported, a parameters pass is still needed. So this is only for Safari > 14.1 and `bugfixes: false`, where arrow function is supported perfectly.

[Babel](https://babeljs.io/repl#?browsers=safari%2014.1&build=&builtIns=false&corejs=3.6&spec=true&loose=false&code_lz=MYGwhgzhAECC0G8BQ1pmgXmgRmgamgApCA6MsASkwD40KVoAjTIskxqjW5ggFwAsAljAYReYXoODRpWQrJwAmAMyda0vkJEBfIA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=true&timeTravel=false&sourceType=module&lineWrap=true&presets=env&prettier=false&targets=&version=7.17.2&externalPlugins=&assumptions=%7B%7D) behaves the same(you need to mannually close bugfixes), and there's no other way to capture `this` in class field

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/src/es2015/parameters.rs:568]: But as this pass is applied after arrow, the emitted code will not work in browsers that only support es5.
I'm not talking about browers with support for arrows.

Comment (Austaras) [reply crates/swc_ecma_transforms_compat/src/es2015/parameters.rs:568]: For browser that doesn't support arrows, there's no chance they support class field, so this shall still work, but implicit dependency between passes is indeed a bad thing.

Comment (kdy1) [reply crates/swc_ecma_transforms_compat/src/es2015/parameters.rs:568]: Oh, I see.
My bad, sorry. I tried creating some example and found that class properties will not exist in ast in such cases. Thanks!

**Review by kdy1 — APPROVED** [review]

**Pull request merged**